File: packages/typescript/integrations/discovery/src/bridge-server.ts
import WebSocket, { WebSocketServer } from "ws";
import type { Bridge, BridgeProvider, RelayHandler } from "./bridge-client";
import { parseBridgeProvider } from "./bridge-client";

export const DEFAULT_BRIDGE_HOST = "127.0.0.1";
export const DEFAULT_BRIDGE_PORT = 9339;
export const DEFAULT_BRIDGE_PATH = "/slop-bridge";

type Logger = { info: (...args: any[]) => void; error: (...args: any[]) => void };

export interface BridgeServerOptions {
  logger?: Logger;
  host?: string;
  port?: number;
  path?: string;
}

/**
 * Bridge WebSocket server — hosts the extension bridge at ws://127.0.0.1:9339/slop-bridge.
 *
 * Mirrors the Go bridge server in apps/cli/bridge/server.go.
 * Accepts connections from the browser extension and other consumers,
 * tracks provider announcements, and relays SLOP messages for postMessage providers.
 */
export function createBridgeServer(
  optionsOrLogger?: BridgeServerOptions | Logger,
): Bridge & { start(): Promise<void> } {
  const {
    logger: log,
    host,
    port,
    path,
  } = normalizeOptions(optionsOrLogger);

  let wss: WebSocketServer | null = null;
  let isRunning = false;
  let changeCallback: (() => void) | null = null;

  const sinks = new Set<WebSocket>();
  const providerMap = new Map<string, BridgeProvider>();
  const relaySubscribers = new Map<string, RelayHandler[]>();

  // --- Broadcasting ---

  function broadcast(msg: Record<string, unknown>) {
    const text = JSON.stringify(msg);
    for (const sink of sinks) {
      if (sink.readyState === WebSocket.OPEN) {
        sink.send(text);
      }
    }
  }

  function replayProviders(ws: WebSocket) {
    for (const bp of providerMap.values()) {
      const msg = {
        type: "provider-available",
        tabId: bp.tabId,
        providerKey: bp.providerKey,
        provider: {
          id: bp.id,
          name: bp.name,
          transport: bp.transport,
          ...(bp.url ? { url: bp.url } : {}),
        },
      };
      ws.send(JSON.stringify(msg));
    }
  }

  // --- Message handling ---

  function handleMessage(msg: Record<string, unknown>) {
    const type = msg.type as string;

    switch (type) {
      case "provider-available": {
        const bp = parseBridgeProvider(msg);
        if (!bp) return;
        providerMap.set(bp.providerKey, bp);
        log.info(`[slop-bridge] Provider available: ${bp.name} (${bp.providerKey})`);
        broadcast(msg);
        changeCallback?.();
        break;
      }

      case "provider-unavailable": {
        const providerKey = msg.providerKey as string;
        if (!providerKey) return;
        providerMap.delete(providerKey);
        relaySubscribers.delete(providerKey);
        log.info(`[slop-bridge] Provider unavailable: ${providerKey}`);
        broadcast(msg);
        changeCallback?.();
        break;
      }

      case "slop-relay": {
        const providerKey = msg.providerKey as string;
        const message = msg.message as Record<string, unknown>;
        if (!providerKey || !message) return;
        // Dispatch to local relay subscribers
        const subs = relaySubscribers.get(providerKey);
        if (subs) {
          for (const handler of subs) handler(message);
        }
        // Rebroadcast to all sinks (extension receives consumer messages,
        // consumers receive extension responses)
        broadcast(msg);
        break;
      }

      case "relay-open":
      case "relay-close": {
        // Forward to all sinks so the extension receives relay control
        broadcast(msg);
        break;
      }
    }
  }

  return {
    /** Bind to port 9339. Resolves when listening, rejects if port is in use. */
    start(): Promise<void> {
      return new Promise((resolve, reject) => {
        wss = new WebSocketServer({
          host,
          port,
          path,
        });

        wss.on("listening", () => {
          isRunning = true;
          log.info(`[slop-bridge] Bridge server running on ws://${host}:${port}${path}`);
          resolve();
        });

        wss.on("error", (err: Error) => {
          if (!isRunning) {
            reject(err);
          } else {
            log.error("[slop-bridge] Server error:", err.message);
          }
        });

        wss.on("connection", (ws: WebSocket) => {
          sinks.add(ws);
          log.info(`[slop-bridge] Client connected (${sinks.size} total)`);

          // Replay current providers to new connection
          replayProviders(ws);

          ws.on("message", (data: WebSocket.Data) => {
            try {
              const msg = JSON.parse(data.toString());
              handleMessage(msg);
            } catch {}
          });

          ws.on("close", () => {
            sinks.delete(ws);
            log.info(`[slop-bridge] Client disconnected (${sinks.size} remaining)`);

            if (sinks.size === 0) {
              // No clients left — clear all relay subscribers and providers
              relaySubscribers.clear();
              providerMap.clear();
              changeCallback?.();
            }
          });
        });
      });
    },

    running() {
      return isRunning;
    },

    providers() {
      return Array.from(providerMap.values());
    },

    onProviderChange(fn: () => void) {
      changeCallback = fn;
    },

    subscribeRelay(providerKey: string): RelayHandler[] {
      const handlers = relaySubscribers.get(providerKey) ?? [];
      relaySubscribers.set(providerKey, handlers);
      return handlers;
    },

    unsubscribeRelay(providerKey: string, handler: RelayHandler) {
      const subs = relaySubscribers.get(providerKey);
      if (!subs) return;
      const idx = subs.indexOf(handler);
      if (idx >= 0) subs.splice(idx, 1);
      if (subs.length === 0) relaySubscribers.delete(providerKey);
    },

    send(msg: Record<string, unknown>) {
      broadcast(msg);
    },

    stop() {
      isRunning = false;
      for (const sink of sinks) sink.close();
      sinks.clear();
      providerMap.clear();
      relaySubscribers.clear();
      if (wss) {
        wss.close();
        wss = null;
      }
    },
  };
}

function normalizeOptions(optionsOrLogger?: BridgeServerOptions | Logger): Required<BridgeServerOptions> {
  const options =
    optionsOrLogger && ("host" in optionsOrLogger || "port" in optionsOrLogger || "path" in optionsOrLogger || "logger" in optionsOrLogger)
      ? optionsOrLogger as BridgeServerOptions
      : { logger: optionsOrLogger as Logger | undefined };

  return {
    logger: options.logger ?? { info: console.error, error: console.error },
    host: options.host ?? DEFAULT_BRIDGE_HOST,
    port: options.port ?? DEFAULT_BRIDGE_PORT,
    path: options.path ?? DEFAULT_BRIDGE_PATH,
  };
}

File: packages/go/slop-ai/discovery/bridge_server.go
package discovery

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"sync"

	"nhooyr.io/websocket"
)

type wsSink struct {
	conn *websocket.Conn
	mu   sync.Mutex
}

func (s *wsSink) write(ctx context.Context, payload []byte) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.conn.Write(ctx, websocket.MessageText, payload)
}

// BridgeServer hosts the extension bridge and mirrors bridge providers for local consumers.
type BridgeServer struct {
	addr   string
	path   string
	logger Logger

	mu               sync.RWMutex
	providers        map[string]BridgeProvider
	relaySubscribers map[string][]chan map[string]any
	sinks            map[*wsSink]struct{}
	httpServer       *http.Server
	listener         net.Listener
	actualAddr       string
	running          bool
	onProviderChange func()
	cancel           context.CancelFunc
	ctx              context.Context
}

// NewBridgeServer creates a new bridge server.
func NewBridgeServer(addr, path string, logger Logger) *BridgeServer {
	return &BridgeServer{
		addr:             addr,
		path:             path,
		logger:           logger,
		providers:        map[string]BridgeProvider{},
		relaySubscribers: map[string][]chan map[string]any{},
		sinks:            map[*wsSink]struct{}{},
	}
}

// Start binds the bridge server and begins serving in the background.
func (s *BridgeServer) Start(ctx context.Context) error {
	mux := http.NewServeMux()
	mux.HandleFunc(s.path, s.handleUpgrade)

	listener, err := net.Listen("tcp", s.addr)
	if err != nil {
		return err
	}

	serverCtx, cancel := context.WithCancel(ctx)
	httpServer := &http.Server{Handler: mux}

	s.mu.Lock()
	s.listener = listener
	s.actualAddr = listener.Addr().String()
	s.ctx = serverCtx
	s.cancel = cancel
	s.httpServer = httpServer
	s.running = true
	s.mu.Unlock()

	go func() {
		<-serverCtx.Done()
		_ = httpServer.Close()
	}()

	go func() {
		err := httpServer.Serve(listener)
		if err != nil && err != http.ErrServerClosed {
			s.logger.errorf("[slop-bridge] Server error: %v", err)
		}
	}()

	s.logger.infof("[slop-bridge] Bridge server running on %s", s.URL())
	return nil
}

// URL returns the current bridge URL.
func (s *BridgeServer) URL() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	addr := s.actualAddr
	if addr == "" {
		addr = s.addr
	}
	return fmt.Sprintf("ws://%s%s", addr, s.path)
}

func (s *BridgeServer) Stop() {
	s.mu.Lock()
	if s.cancel != nil {
		s.cancel()
	}
	listener := s.listener
	httpServer := s.httpServer
	s.running = false
	s.listener = nil
	s.httpServer = nil
	providersChanged := len(s.providers) > 0
	for key, subs := range s.relaySubscribers {
		for _, ch := range subs {
			close(ch)
		}
		delete(s.relaySubscribers, key)
	}
	s.providers = map[string]BridgeProvider{}
	sinks := make([]*wsSink, 0, len(s.sinks))
	for sink := range s.sinks {
		sinks = append(sinks, sink)
		delete(s.sinks, sink)
	}
	onChange := s.onProviderChange
	s.mu.Unlock()

	if listener != nil {
		_ = listener.Close()
	}
	if httpServer != nil {
		_ = httpServer.Close()
	}
	for _, sink := range sinks {
		_ = sink.conn.Close(websocket.StatusNormalClosure, "")
	}
	if providersChanged && onChange != nil {
		onChange()
	}
}

func (s *BridgeServer) Running() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.running
}

func (s *BridgeServer) Providers() []BridgeProvider {
	s.mu.RLock()
	defer s.mu.RUnlock()
	providers := make([]BridgeProvider, 0, len(s.providers))
	for _, provider := range s.providers {
		providers = append(providers, provider)
	}
	return providers
}

func (s *BridgeServer) OnProviderChange(fn func()) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.onProviderChange = fn
}

func (s *BridgeServer) SubscribeRelay(providerKey string) chan map[string]any {
	ch := make(chan map[string]any, 64)
	s.mu.Lock()
	s.relaySubscribers[providerKey] = append(s.relaySubscribers[providerKey], ch)
	s.mu.Unlock()
	return ch
}

func (s *BridgeServer) UnsubscribeRelay(providerKey string, ch chan map[string]any) {
	s.mu.Lock()
	defer s.mu.Unlock()
	subs := s.relaySubscribers[providerKey]
	for index, sub := range subs {
		if sub == ch {
			s.relaySubscribers[providerKey] = append(subs[:index], subs[index+1:]...)
			break
		}
	}
	if len(s.relaySubscribers[providerKey]) == 0 {
		delete(s.relaySubscribers, providerKey)
	}
	close(ch)
}

func (s *BridgeServer) Send(msg map[string]any) error {
	return s.broadcast(msg)
}

func (s *BridgeServer) broadcast(msg map[string]any) error {
	payload, err := json.Marshal(msg)
	if err != nil {
		return err
	}

	s.mu.RLock()
	ctx := s.ctx
	sinks := make([]*wsSink, 0, len(s.sinks))
	for sink := range s.sinks {
		sinks = append(sinks, sink)
	}
	s.mu.RUnlock()
	if ctx == nil {
		ctx = context.Background()
	}

	for _, sink := range sinks {
		if err := sink.write(ctx, payload); err != nil {
			s.logger.errorf("[slop-bridge] Broadcast failed: %v", err)
		}
	}
	return nil
}

func (s *BridgeServer) handleUpgrade(w http.ResponseWriter, r *http.Request) {
	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{InsecureSkipVerify: true})
	if err != nil {
		return
	}

	sink := &wsSink{conn: conn}
	s.addConn(sink)
	defer s.removeConn(sink)

	s.readLoop(sink)
}

func (s *BridgeServer) addConn(sink *wsSink) {
	s.mu.Lock()
	s.sinks[sink] = struct{}{}
	providers := make([]BridgeProvider, 0, len(s.providers))
	for _, provider := range s.providers {
		providers = append(providers, provider)
	}
	ctx := s.ctx
	s.mu.Unlock()

	if ctx == nil {
		ctx = context.Background()
	}

	for _, provider := range providers {
		message := map[string]any{
			"type":        "provider-available",
			"tabId":       provider.TabID,
			"providerKey": provider.ProviderKey,
			"provider": map[string]any{
				"id":        provider.ID,
				"name":      provider.Name,
				"transport": provider.Transport,
			},
		}
		if provider.URL != "" {
			message["provider"].(map[string]any)["url"] = provider.URL
		}
		payload, err := json.Marshal(message)
		if err != nil {
			continue
		}
		_ = sink.write(ctx, payload)
	}
}

func (s *BridgeServer) removeConn(sink *wsSink) {
	s.mu.Lock()
	delete(s.sinks, sink)
	noSinks := len(s.sinks) == 0
	providersChanged := false
	if noSinks {
		if len(s.providers) > 0 {
			providersChanged = true
		}
		for key, subs := range s.relaySubscribers {
			for _, ch := range subs {
				close(ch)
			}
			delete(s.relaySubscribers, key)
		}
		s.providers = map[string]BridgeProvider{}
	}
	onChange := s.onProviderChange
	s.mu.Unlock()

	_ = sink.conn.Close(websocket.StatusNormalClosure, "")
	if providersChanged && onChange != nil {
		onChange()
	}
}

func (s *BridgeServer) readLoop(sink *wsSink) {
	for {
		s.mu.RLock()
		ctx := s.ctx
		s.mu.RUnlock()
		if ctx == nil {
			ctx = context.Background()
		}

		_, data, err := sink.conn.Read(ctx)
		if err != nil {
			return
		}

		var msg map[string]any
		if err := json.Unmarshal(data, &msg); err != nil {
			continue
		}
		s.handleMessage(msg)
	}
}

func (s *BridgeServer) handleMessage(msg map[string]any) {
	msgType, _ := msg["type"].(string)

	switch msgType {
	case "provider-available":
		provider, ok := parseBridgeProvider(msg)
		if !ok {
			return
		}
		s.mu.Lock()
		s.providers[provider.ProviderKey] = provider
		onChange := s.onProviderChange
		s.mu.Unlock()
		_ = s.broadcast(msg)
		if onChange != nil {
			onChange()
		}

	case "provider-unavailable":
		providerKey, _ := msg["providerKey"].(string)
		if providerKey == "" {
			return
		}
		s.mu.Lock()
		delete(s.providers, providerKey)
		if subs, ok := s.relaySubscribers[providerKey]; ok {
			for _, ch := range subs {
				close(ch)
			}
			delete(s.relaySubscribers, providerKey)
		}
		onChange := s.onProviderChange
		s.mu.Unlock()
		_ = s.broadcast(msg)
		if onChange != nil {
			onChange()
		}

	case "slop-relay":
		providerKey, _ := msg["providerKey"].(string)
		message, _ := msg["message"].(map[string]any)
		if providerKey == "" || message == nil {
			return
		}
		s.dispatchRelay(providerKey, message)
		_ = s.broadcast(msg)

	case "relay-open", "relay-close":
		_ = s.broadcast(msg)
	}
}

func (s *BridgeServer) dispatchRelay(providerKey string, message map[string]any) {
	s.mu.RLock()
	subs := append([]chan map[string]any(nil), s.relaySubscribers[providerKey]...)
	s.mu.RUnlock()

	for _, ch := range subs {
		select {
		case ch <- message:
		default:
		}
	}
}
